pennylane/_grad.py

```diff
@@ -181,3 +182,206 @@ def _jacobian_function(*args, **kwargs):
         return _np.stack([_jacobian(func, arg)(*args, **kwargs) for arg in argnum]).T
 
     return _jacobian_function
+
+
+def _fd_first_order_centered(f, argnum, delta, *args, idx=None, **kwargs):
+
+    r"""Uses a central finite difference approximation to compute the gradient
+    of the function ``f`` with respect to the argument ``argnum``.
+
+    Args:
+        f (function): function with signature ``f(*args, **kwargs)``
+        argnum (int): the argument with respect to which the gradient is taken
+        delta (float): step size used to evaluate the finite difference
+        idx (list[int]): if argument ``args[argnum]`` is an array, ``idx`` can
+            be used to specify the indices of the arguments to differentiate
+
+    Returns:
+        (float or array): the gradient of the input function with respect
+        to the arguments in ``argnum``
+    """
+
+    if argnum > len(args) - 1:
+        raise ValueError(
+            "The value of 'argnum' has to be between 0 and {}; got {}".format(len(args) - 1, argnum)
+        )
+
+    x = _np.array(args[argnum])
+    gradient = _np.zeros_like(x, dtype="O")
```

Comment:
Should we make this a list, `gradient = []`?

Then, further down, we can do

```python
        gradient.append(
            (f(*args[:argnum], x + shift, *args[argnum + 1 :], **kwargs)
            - f(*args[:argnum], x - shift, *args[argnum + 1 :], **kwargs)) * delta ** -1
        )

    return _np.array(gradient).reshape(x.shape)
```

I _think_ this might be beneficial, in that we get a float array or an object array when needed, rather than it _always_ being an object array. @agran2018 maybe worth changing, and seeing if the tests still pass?

Reply:
@josh146, I tried this yesterday and it did not work for me:

```
   247     # return gradient
--> 248     return _np.array(gradient).reshape(x.shape)
    249 
    250 

ValueError: cannot reshape array of size 4 into shape (2,)
```
Also, we need to keep track of the indices corresponding to the zero and non-zero components of the gradients if the list ``idx`` is given. This is important for constrained geometry optimization which is very useful.

Reply:
Oh, of course - the shape output should be the output shape of `f` 🤦